src/keyboard.rs

```diff
@@ -185,26 +185,112 @@ impl std::fmt::Debug for NativeKey {
     }
 }
 
+impl From<NativeKeyCode> for NativeKey {
+    #[inline]
+    fn from(code: NativeKeyCode) -> Self {
+        match code {
+            NativeKeyCode::Unidentified => NativeKey::Unidentified,
+            NativeKeyCode::Android(x) => NativeKey::Android(x),
+            NativeKeyCode::MacOS(x) => NativeKey::MacOS(x),
+            NativeKeyCode::Windows(x) => NativeKey::Windows(x),
+            NativeKeyCode::Xkb(x) => NativeKey::Xkb(x),
+        }
+    }
+}
+
+impl PartialEq<NativeKey> for NativeKeyCode {
+    #[allow(clippy::cmp_owned)] // uses less code than direct match; target is stack allocated
+    #[inline]
+    fn eq(&self, rhs: &NativeKey) -> bool {
+        NativeKey::from(*self) == *rhs
+    }
+}
+
+impl PartialEq<NativeKeyCode> for NativeKey {
+    #[inline]
+    fn eq(&self, rhs: &NativeKeyCode) -> bool {
+        rhs == self
+    }
+}
```

Comment:
No we can't `derive`: these are cross-type comparisons. (We don't need them either, but I consider them slightly useful and not harmful.)

Reply:
I mean, do we really need them? Given that we can `from` and then compare. Cross type compare is a rare in rust.

Reply:
I've found this type of comparison useful before and added one or two token usages in the examples... but I think the only common usage in Rust is comparing things like [`String` to `&str`](https://doc.rust-lang.org/stable/std/string/struct.String.html#impl-PartialEq%3C%26'a+str%3E-for-String) or `String` to `Cow<str>` (both `Deref` to the same type but `Deref` alone won't make the comparison work).

At any rate, I don't think these comparisons are dangerous (mis-usable). But I can remove if you don't want them?

Reply:
I'd prefer to remove them just for forward compat and because we have `into`.

Reply:
I actually find these useful.

> I'd prefer to remove them just for forward compat and because we have `into`.

What do you have in mind with "forward compat"? It is true that these comparisons could be made with `into()`, but if it's a useful improvement with no downsides, I would still be all for it.

Reply:
I mean that in the future we could end in a state where we can't really compare them anymore. I still like the `into`, since it's how you do even with `ints` in rust. We could implement `as` if it matters.

Reply:
> I mean that in the future we could end in a state where we can't really compare them anymore.

I mean do we care about that future? We can always remove them if it becomes incomparable, we aren't at v1 yet.

> We could implement `as` if it matters.

Do you mean `Deref`? I think that also would be a nice addition, then comparisons could be made without this implementation as well, if it all works out that is.

Reply:
Not `Deref`, but `as_physical`, but it doesn't make any sense, since the type is `Copy`.

Reply:
https://github.com/rust-windowing/winit/pull/3143#discussion_r1360324557 just saw that `Deref` was already discussed/mentioned here, sorry.

Yeah, that wouldn't really make sense to me either. I don't really see the downside of adding these `PartialEq` implementations, so I'm still in favor.

Reply:
I don't think `Deref` can be implemented for any of these.